### ISSUE TITLE
100% cpu usage

### ISSUE THREAD

**Issue by jamesmarlowe** [pintsized/lua-resty-http#37]
Hi, I'm trying to use this library to send json encoded data (cjson.encode()) to an external website and it seems to work a few times (read 10s) but then the cpu usage goes to 100% and locks up until I kill nginx. I set a timeout and a keepalive and if I set the request body to something fake, even small json snippets it works fine.

I'm not sure if I even know what to ask to try to get somewhere with it since it seems like a very strange problem. Is there some case where the request body is tried to be interpreted in a way that would infinitely loop if the data were parsable? or perhaps if it finds a url or ip in the body would try to call them maybe? Or a max character limit that would loop trying indefinitely to send the next buffer size? I realize these are long shots, but I can't figure out what is happening.

If there is an issue found in body parsing, I'd be happy to contribute, but I can't figure out what is making this happen, any help appreciated.

**Comment by pintsized** [comment]
Hi, can you post a minimal example to recreate this with some example data? It's a bit hard to guess what's going on otherwise. I'll try and take a look later this evening.

James.

**Comment by jamesmarlowe** [comment]
Thanks for your quick reply, I've been trying to narrow it down, but haven't succeeded yet. I'll try to post something smallish that displays the behavior later today

**Comment by hamishforbes** [comment]
If the body param is a function then it will be called in a loop until it returns `nil` https://github.com/pintsized/lua-resty-http/blob/master/lib/resty/http.lua#L462
Are you possibly setting the body param to be `cjson.encode` rather than the *return* of `cjson.encode()`?

Thats the only obvious way I can think of that the request would hang forever and be consuming CPU.
I think any other loops would be sitting waiting for network I/O which shouldn't consume CPU.

**Comment by jamesmarlowe** [comment]
Ehhh sorry guys, I found the problem and it wasn't this library at all, it was behaving perfectly. It was in parsing the response afterward and I incorrectly thought it was here because sending different bodies through elicited a different response from the other end which would change the parsing behavior which I bundled together with the requesting behavior.

Testing for sending the function cjson.encode was a great idea, I ended up sending `body = request_body..""` just to make sure it was actually a string as I thought.
